### PR TITLE
Replace header guards with #pragma once

### PR DESCRIPTION
**Pull request by lmarz**
Signed-off-by: Leon Marz <main@lmarz.org>

Part of: #508 

First of all, sorry for the huge PR. I can split it up into multiple commits if you think it's better.
I renamed every header guard in the src directory in the style `__TENACITY_FILE_NAME__`, except for files with Audacity in their name and the files in the `include/tenacity` directory. There I used the style `__TENACITY_FILE_NAME_H__`.
I might have missed some header files, since there wasn't really a standard convention and I had to manually check every file.
For example the ffmpeg header files had `#if !defined(` and some headers didn't even have a header guard.

<details>
<summary>Checklist</summary>

- [x] I have signed off my commits using `-s` or `Signed-off-by`\* (See: [Contributing § DCO](https://github.com/tenacityteam/tenacity/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code\*
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving\*
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"\*

\* indicates required

</details>

### CODE REVIEW

**Review by nbsp — COMMENTED** [review]
A couple of headers are still marked `__AUDACITY_*`. I think it's fine to change those even if the filenames have Audacity in them. We don't change that purely for upstream compatibility reasons.

**Comment by nbsp** [comment]
As for the commit size: don't worry, it's a perfectly sized commit.

**Comment by Be-ing** [comment]
I would prefer to replace these all with the less error prone `#pragma once`, but I don't really care.

**Comment by lmarz** [comment]
> I would prefer to replace these all with the less error prone `#pragma once`, but I don't really care.

`#pragma once` is not in the c++ standard, although it is widely supported by compilers. Apart from possible conflicts when there are files with the same name, it mostly comes down to personal preference.  Audacity almost entirely uses header guards and I personally like it more

File: src/AudacityFileConfig.h [hunk 1]
```diff
@@ -9,8 +9,8 @@ Paul Licameli split from Prefs.h
 
 **********************************************************************/
 
-#ifndef __AUDACITY_FILE_CONFIG__
-#define __AUDACITY_FILE_CONFIG__
+#ifndef __TENACITY_AUDACITY_FILE_CONFIG__
+#define __TENACITY_AUDACITY_FILE_CONFIG__
```
Comment (nbsp) [inline src/AudacityFileConfig.h]: Hmm, I'm not sure I like how this looks.

Comment (lmarz) [reply src/AudacityFileConfig.h]: There is a `FileConfig.h` in src/widgets which already has `__TENACITY_FILE_CONFIG__`

Comment (nbsp) [reply src/AudacityFileConfig.h]: Oh. Audacity dealt with it by adding `WIDGETS`, which is a more elegant solution imo.

Comment (lmarz) [reply src/AudacityFileConfig.h]: Agreed

Comment (nbsp) [reply src/AudacityFileConfig.h]: Now, do we want to only add that for things that might overlap? Or is it better to add `WIDGETS` to everything under src/widgets/?

**Comment by leio** [comment]
My opinion is that we should verify that `#pragma once` works everywhere we need, and use it instead of the guard naming mess. In the C world at least it's universally available and not really a question if it can be used or not, at least for GUI applications.

**Review by Be-ing — APPROVED** [review]

**Review by n0toose — APPROVED** [review]
Looks good to me!

**Comment by n0toose** [comment]
Okay, let's orient ourselves and use this thread to argue the pros and cons of #pragma once and the re-merge this change if we deem it fit for it to stay. (Apologies in advance to @lmarz, I understand that this could possibly be very frustrating to you.)

**Comment by emabrey** [comment]
I have reverted this commit because there was some confusion about how long to wait before an approved PR can be merged. There had been discussion about this but we hadn't set the rules into a concrete document, but that clearly isn't working. We wanted to make sure, due to the fact that contributors are in multiple time zones etc., that everyone had a reasonable chance to give comments and reviews to any PRs **before** they are merged. I just want to make sure that the contributor to this PR knows that I didn't revert it because of anything to do with their contribution and that after the minimum time has elapsed, it can still be merged. Sorry for any confusion this may have caused, and once again I want to make sure that everyone knows this was not a reflection on the underlying content of the merge.

**Comment by n0toose** [comment]
Pros for implementing `#pragma once`:
- Never caused me issues in other projects.
- The fact that it isn't standardized is basically the point, as each implementation is optimized for the compiler.
- ***All*** major compilers support it, the last time when it was removed from GCC was probably before I was even born.
- Name conflicts are avoided.
- Some scenarios, such as "two files have the exact same content" is something that we should avoid either way. Some other scenarios that some implementations previously dealt with the wrong way only really had to do with scenarios such as "two symbolic links pointing to the same file", which is something we should ***definitely avoid*** either way.
- Name conflicts are much more probable than the two aforementioned scenarios that count as disadvantages.
- `#pragma once` can improve compilation times.

**Comment by n0toose** [comment]
It seems like builds with `#pragma once` are relatively faster based on the CI output, but the margin of error is probably very high here. This needs further testing.

**Comment by Be-ing** [comment]
The variability of build times on GitHub Actions is pretty high.

**Comment by n0toose** [comment]
> The variability of build times on GitHub Actions is pretty high.

Yup. I wanted to just emphasize that they are not a logically sound way to reach a conclusion.

**Comment by emabrey** [comment]
MSVC [already does the multiple include optimization for header guards](https://docs.microsoft.com/en-us/cpp/preprocessor/once?redirectedfrom=MSDN&view=msvc-160): "The compiler recognizes the include guard idiom, and implements the multiple-include optimization the same way as the #pragma once directive if no non-comment code or preprocessor directive comes before or after the standard form of the idiom[...]" (they mean that you can't have code outside the guard, it's worded poorly).

GCC [already does the same optimizations](https://gcc.gnu.org/onlinedocs/cppinternals/Guard-Macros.html) for header guards: 

"If the file were included a second time, it can only be optimized away if that inclusion would result in no tokens to return, and no relevant directives to process." (meaning it similar restrictions to MSVC about no intervening preprocessor defines etc.)

Since GCC does it, so does Clang.

I would also have concerns about how it interacts with ccache and sccache: https://ccache.dev/manual/latest.html#_how_ccache_works

If you want some real world performance results comparing the two approaches (guards versus pragma), here they are: https://github.com/Return-To-The-Roots/s25client/issues/1073#issuecomment-508055696

Therefore, based on those results, you can see that`#pragma once` does not offer meaningful performance benefits, and in fact, it can actually be meaningfully slower on some configurations/compilers. Additionally, it also is non-standardized, meaning compilers can do different things with it. Additionally, where the `#ifndef` approach allows for the handling of unusual build scenarios, `#pragma once` does not. On MSVC the `#pragma once` solution is approximately 50% faster, and on GCC it is approximately 667% _slower_ (meaning about 15% as fast); when you consider the fact that this forced optimization isn't even always optimal, I struggle to see a justification for it.

Really the only benefit to `#pragma once` is that it saves you about 2 lines of boilerplate a file. The downside is that it can break the build in ways header guards cannot. When using `#pragma once` you are tying the compiler's hands and forcing it to do it an optimization, or fail. With header guards if it cannot do the optimization it just does the slow path and the build still works.

**Comment by Be-ing** [comment]
> Really the only benefit to #pragma once is that it saves you about 2 lines of boilerplate a file.

It is not just that it saves boilerplate, but that 3 lines of boilerplate is error-prone and spread across the file. I have had build failures due to typos or missing `#endif`s with header guards. On the other hand, I have never had a problem with `#pragma once`.

**Comment by emabrey** [comment]
Do you not consider potentially taking about 6 times as long to compile a problem with `#pragma once`? Also it only saves 2 lines of boilerplate, not 3. The `#pragma once` directive is itself a line of boilerplate.

**Comment by Be-ing** [comment]
No, not really. I've not noticed any difference in build times. Maybe there is one if you measure it. But I suspect the total difference in build times would be practically insignificant. In the real world, I am much more interested in preventing errors from typos.

**Comment by emabrey** [comment]
The plural of anecdote is not data. Saying "I have never had an issue" is about as useful as "it compiles on my machine". If we are making this change to reduce risk of typos then this is not the appropriate solution to that problem. That is what static analysis is designed for, not compiler directives.

I'm not refuting that the risk of typos is a problem, but I would contest that the `#pragma` directives are less prone to typos. If you type `#pragma oncee` the compiler isn't able to know if it just doesn't support that directive, so it can't meaningfully do anything other than ignore it (and maybe issue a warning if it is configured to do so). You could easily have a scenario where that typo doesn't immediately break until sometime later when say a template or something else that is required to have a single definition is put into the file. So regarding typographical errors, compiler directives are just as prone as header guards, though in a different way.

I think this is more of a "spaces" vs. "tabs" dispute where there is not a meaningful programmatic reason to differentiate them, and it boils down to a discussion about code style. If you want to prevent typos and enforce coding style, using a compiler directive is absolutely a screwdriver solution to a hammer problem. Header guards are provably better than compiler directives for managing inclusion, which is why I have provided data to prove that. If we want to avoid typos we should add a standardized way to generate header guard names from files and make it an enforceable static analysis check. That would be in scope for the discussions we have already had about code style and how and what kind of enforcement we want that code style to take.

**Comment by n0toose** [comment]
Anything that could make contributions easier and the entire project more approachable seems like a good idea to me. I could ask other software projects about their use of `#pragma once` and the ways they justify it. It's probably not just me if big names like [FreeBSD](https://github.com/freebsd/freebsd-src/search?q=pragma+once), [OpenSSL](https://github.com/openssl/openssl/search?q=pragma+once) and [curl](https://github.com/curl/curl/search?q=pragma+once) use `#pragma` headers in the real world.

**Comment by n0toose** [comment]
[Wikipedia's table regarding portability is also pretty nice.](https://en.wikipedia.org/wiki/Pragma_once#Portability)

`#pragma once` being removed once probably before I was even born from one specific compiler does not sound convincing enough to me, as far as the potential of any compiler being able to remove it at some point. As far as the `#pragma once` header is concerned, the fact that it's not standard is better.

**Comment by emabrey** [comment]
Please address the concerns I have brought up rather than continue to use _argumentum ad populum_. Additionally, the FreeBSD search results you provided utilized an incorrect search string. When using the correct search string I found 122 instances of `#pragma once` in what I believe are about 87,333 files (I'm not sure how to get the file count on Github so that was just using a \*.\* search string, so it may be off). That means FreeBSD uses it at a 0.1% rate, which is hardly a ringing endorsement and more of a rounding error. OpenSSL appeared to have 243 usages in 24,225 files, or 1%. As far as curl, I could not find any instances of `#pragma once`: https://github.com/curl/curl/search?q=%22%23pragma+once%22

Edit: I did find this curl issue that uses it though, https://github.com/curl/curl/issues/2421, but it was not accepted.

**Comment by emabrey** [comment]
Actually, limiting the FreeBSD results to only C and C++ files results in a usage rate of 0.3%, not 0.1%.

**Comment by emabrey** [comment]
Also, the concerns I have about `#pragma once` extends outside the scope of which compiler's do and do not support it. The correct implementation of the directive is actually the problem I have with it. I am conceding that it does have far reaching support, but that still doesn't address the crux of the problem. You are not wrong that all the compilers we are likely to encounter support it's usage at the current time. Additionally you are correct that they probably won't remove support anytime soon.

The problem with `#pragma once` is that it acts like this (hypothetical psuedo-code for explaining my point):
```
try{
    optimize_or_error()
} catch{ 
    exit(1); //uh-oh we broke the build
}
```

whereas header guards act like this:
```
if(canOptimize){
    optimize_or_error() // never errors because we already checked
} else{
    do_include_slowly() //never errors because it does things methodically
}
```

There is basically no performance benefit to `#pragma once` on any modern compiler that isn't within the margin-of-error. The only real justification to use it is that you are more worried about typos in header guards than typos in compiler directives. Even that is barely a justification, considering that you can use static analysis. Here is an example (admittedly an older example):

https://fuchsia.googlesource.com/scripts/+/8acbd7a7f9ef1459fa43e8c168dd2091d6fe2477/check-header-guards.py

The Google code style guide discourages usage of `#pragma once`: https://github.com/google/styleguide/issues/538 (meaning so does https://github.com/cpplint/cpplint)

There are standardized tools for conversion to and from these style choices, and also to verify the include header guards:

https://github.com/cgmb/guardonce

**Comment by n0toose** [comment]
Yeah, it's very true that compilers aren't as dumb anymore and that speed as a result of bad or just idiosyncratic code is not as much of an issue nowadays...

Should've probably conducted research from a more critical point of view.

**Comment by Be-ing** [comment]
I don't know if continuing to discuss this is going to change anyone's mind. Shall we vote?

**Comment by emabrey** [comment]
I believe that if we are going to vote on this it should be on an entire code style, not just this one singular component of a comprehensive code style. It would be best to pick a code style that allows `#pragma once` if that is what we decide to do, so if we decide now to implement it then we are limiting future code styles we could choose without defining one from scratch (which is annoying because writing config files for static analysis tools isn't fun). I was incorrect about cpplint earlier, it does allow `#pragma once`  even though the official google style guide doesn't.

So give me a few minutes and I will make a table of popular code styles and if they support `#pragma once` and we can go from there if it's actually a limiting factor (maybe the google style is unique in this regard).

**Comment by n0toose** [comment]
I try to use [Haiku's Coding Guidelines](https://www.haiku-os.org/development/coding-guidelines/) as a point of reference sometimes, when I am not sure how to indent things properly. Mostly C-style C++ code, like what we have here. No mention of `#pragma once`, though.

**Comment by Be-ing** [comment]
I'd rather not get into a larger discussion about an entire code style guide right now. For now we need to get rid of the references to Audacity from the code.

**Comment by emabrey** [comment]
| Style Name | Link to Style | Supports `#pragma once` |
|---|---|---|
| Google (Official Google) | https://google.github.io/styleguide/cppguide.html | No |
| cpplint (Unofficial Google) | https://github.com/cpplint/cpplint | Yes, optional |
| c++ Core (Official C++ Standard) | https://github.com/isocpp/CppCoreGuidelines | No |
| Unreal | https://docs.unrealengine.com/4.27/en-US/ProductionPipelines/DevelopmentSetup/CodingStandard/ | Yes, required |
| LLVM  | https://llvm.org/docs/CodingStandards.html | No |
| Mozilla | https://firefox-source-docs.mozilla.org/code-quality/coding-style/coding_style_cpp.html | No, except legacy |
| WebKit | https://webkit.org/code-style-guidelines/ | Yes, required |
| Qt | https://wiki.qt.io/Qt_Coding_Style | Yes, optional |
| Blender | https://wiki.blender.org/wiki/Style_Guide/C_Cpp | Yes, optional |
| Blink | https://chromium.googlesource.com/chromium/src/+/refs/heads/main/styleguide/c++/blink-c++.md | No |
| Inkscape | https://inkscape.org/develop/coding-style/ | No |
|  |  |  |

**Comment by emabrey** [comment]
I did just find out that there is an official C++ style guide created with the help of Bjarne Stroustrup, which is awesome and I had not heard of until now. It says that [you should not use `#pragma once`](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#note-345) because it violates the [rule P.2](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rp-Cplusplus) of the official C++ code style guide.

**Comment by emabrey** [comment]
Also I didn't include the base Chromium standard because it says it only supports c++14, though I included Blink which seems to be based upon Chromium and which I think supports styling c++17.

**Comment by emabrey** [comment]
To be clear I'm willing to use `#pragma once` or not use it; I just want to have a consistent standard that everyone applies. I obviously believe that the best coding standard is to forbid it's use, but I'm not unaware of the fact that ultimately the coding standard needs to meet the needs of all the contributors, not just what is or is not "a best practice". Personally, knowing the risk of header guard typos, I try to double or triple check them. Some people might not do that though, so they would obviously prefer `#pragma once` since it's more copy/paste-able.

Ultimately I think the real solution is to implement static analysis status checks as part of our merge approval process, and to check for header guard typos using something like https://github.com/cgmb/guardonce. If we don't do that, we should still check for compiler directives and reject any that are not on a whitelist (such as accepting `#pragma once` but not other `#pragma` directives) to catch typos and reject indiscriminate usage of vendor-specific directives.

Ultimately, we should pick a coding style and enforce it. Once we pick a style, we can transition the header guards to `#pragma once` if we want as part of that transition. If we don't use directives, we can enforce guard names match an agreed upon pattern, perhaps based upon the file name or something else entirely, so long as it is reasonably unique. See https://github.com/cgmb/guardonce/blob/master/docs/Walkthrough.md for examples.

**Comment by Be-ing** [comment]
Header guards are also a hassle when renaming files.

**Pull request merged**